### ISSUE TITLE
Compiled EventSubscriber for possible performance optimization

### ISSUE THREAD

**Issue by grisendo** [symfony/symfony#26993]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes
| Symfony version  | x

I was doing some debugging with a custom CMS made with Symfony (3.x) using Blackfire, in order to improve the performance as much as possible. I am not a profiler expert, but I saw that a huge amount of time was spent with "EventDispatcher" class, at "sortListeners" method. So, I was thinking: I always add events listeners/subscribers using tags in service container, so why subscribed classes are not already added and sorted in the container when compiling? Why not to create a CompiledEventSubscriber that you can optionally use instead of normal one, if you are pretty sure you only work with container-tag defined event subscriptions?

I don't know if there is already something about this topic, I was discussing it today at Slack, and @ostrolucky show me his own solution: https://github.com/ostrolucky/app-event-dispatcher/blob/master/src/Symfony/DependencyInjection/RegisterListenersPass.php

**Comment by Koc** [comment]
ref #12521

**Comment by nicolas-grekas** [comment]
I'm not sure there is anything left to optimize. I reviewed @ostrolucky's code but I don't understand how it is related.

`sortListeners` does two things: it sorts, and it initializes listeners that are to be called. Initialization is just required, there is no way around.

IMHO, all this is just legitimate from the pov of `EventDispatcher`, everything is already as optimized as possible AFAIK. You might want to review why you have that many listeners, and what they do in their constructurs.

**Comment by ostrolucky** [comment]
Yeah I guess bottleneck in your case is initializing of listeners, not sorting them. Moving sorting part to compiling would save just few krsort calls, rest of the code would still need to be executed at runtime.

**Comment by nicolas-grekas** [comment]
Closing as explained, please provide additional info if you think this explanation is not the correct one.

**Comment by grisendo** [comment]
Yes, I think the real bottleneck (well, is not such a real-problem big bottleneck, in fact, just me wanting to learn better the Symfony internals) comes from injecting twig/templating service in event listener... not sure 100% yet, debugging in progress. I was just thinking if those ksorts and loop inside loop (and an external loop which executes the function call) could be avoided with container compilation... I am going to check with some tests.

**Comment by grisendo** [comment]
Ok, I saw sortListeners actually does more that its name says, and it instantiates also dependant service classes by executing closures. Then, I delayed all heavy services initialization (in my case doctrine and templating/twig) but I cannot delay translator since TranslatorListener is inside framework bundle (by the way, is there any issue/pull request about delaying this also?).

So here is the result... I am not very used to Blackfire (or any other PHP Profiler) but I don't know where that 6% of difference is wasted (and I don't know how to get that information).

![sweetonion profile blackfire](https://user-images.githubusercontent.com/1665659/39095070-8ccd9c12-463a-11e8-972c-6de7a6b9e189.png)

**Comment by nicolas-grekas** [comment]
If you click on the magnifying glass on the sortListener node you should be able to inspect it with more details. See also the left panel which should contain the other callees.

**Comment by grisendo** [comment]
This is already after clicking a manifier glass, and inside that, no more manifying glass in sortListener node (maybe because doesn't call other methods, just loops and PHP core functions) :disappointed: Anyway, looking at the left panel as you told me I see it takes that amount of exclusive cost

![sweetonion profile blackfire 1](https://user-images.githubusercontent.com/1665659/39099202-cd75c5fa-4676-11e8-84b9-96717409c39f.png)

**Comment by grisendo** [comment]
By the way,  here is a profile i managed to make public for another issue https://blackfire.io/profiles/8e320963-fced-48b8-a7ca-d2f55426f772/graph

**Comment by wadjeroudi** [comment]
@nicolas-grekas I think @grisendo is right and there's still room for improvment for the sortListeners.
On a sylius project, the sortListeners is called 10times in each request, on some pages it can take up to 30% (iwall and ewall).
Why not sort them in the compiling part only one time ?

**Comment by lyrixx** [comment]
I did not read the whole thread but we can not sort them at compile time since it's possible to add new listeners at runtime. Btw symfony itself does that.

**Comment by wadjeroudi** [comment]
it should be sorted by default and symfony should avoid that practice if possible ?
sorting would be fired only for people who really need to add new listeners at runtime.
@lyrixx do you have an example of listener added at runtime by symfony ?

**Comment by grisendo** [comment]
I was thinking about storing a boolean property ($sorted)/array of booleans (for each event type). Then, setting it to "false" when adding a new listener, setting it to "true" after sorting them, and also "true" by default (after compiling the container, where they get initially sorted, as I proposed)... and avoid sorting if true === $sorted

**Comment by nicolas-grekas** [comment]
@grisendo that's already how it works :)

**Comment by lyrixx** [comment]
> @lyrixx do you have an example of listener added at runtime by symfony ?

@wadjeroudi see 
https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/Security/Http/Firewall.php#L71
and 
https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php#L74-L85

**Comment by grisendo** [comment]
@nicolas-grekas many time since I open this issue, I can't remember how it worked :smile: I'll try to check the code again. So, if there's already a/some boolean/s to control about sorting status, it's just about to get them sorted when compiling the container and set that value/s to "true" by default, right?

**Comment by wadjeroudi** [comment]
@nicolas-grekas can we at least remove ksort call ?
and use
```php
$this->sorted[$eventName][$priority] = $listener;
```
will try to investigate more but maybe the execution of the closure can take time, but it could be already there in the old sorted listeners ?

**Comment by wadjeroudi** [comment]
ok, after more digging there's no room for improvment and my propositions were totally wrongs (except for the initialization of the closure, but that's irrelevant).

the only solution would be to insert the listener at the correct order and not reorder the whole sorted array, maybe with something like that : 
https://doc.nette.org/en/2.4/arrays#toc-insertbefore

**Comment by nicolas-grekas** [comment]
For reference, I checked the code again, and everything is already explained in this thread, see the comments I made last year: `ksort()` is FAST. sortListeners does one more thing than calling this function: it initializes the listeners themselves. This MUST be done, of course, so that we can call them. If this step is slow, check your listeners' constructors. Nothing else can be done from the component's side.
